Gemfile

```diff
@@ -91,6 +92,7 @@ gem 'faraday_middleware', '~> 0.13.1'
 gem 'mimemagic', '~> 0.3.10'
 gem 'nokogiri', '~> 1.13.10'
 gem 'secure_headers', '~> 6.3.0'
+gem 'redlock'
```

Comment:
Thank you for all comments!

> Under no circumstances concurrent threads (e.g. concurrent sidekiq jobs of the same process) can share a same redlock client as the value of the token that ensures the lock 100% depends on it.

You mean that if we use same client for multiple threads, if they try to acquire the same lock, then (because the client is still the same) all of them will consider that they have the lock. Do I understand your like of thinking correctly?

> If a lock has to be carried across the stack, rely on (and honour!) the extend: { value: … } options of #lock when instantiating the redlock client. I don't think that's the case as methods should be concise and self-contained.

Do you mean, that in case we need to acuire/extend the lock with another redlock client, we have to pass also the original lock_info, otherwise we get the wrong results?

Agreed that we should not need this.

> I'd check the gem for FIPS compliance. I see digests based on SHA1 there, which should be fine IIRC. But just in case.

Excellent point. Ideally we get rid of SHA1 because it is falling out of favor as well.

Reply:
> You mean that if we use same client for multiple threads […] Do I understand your line of thinking correctly?

Yes. Exactly.

> Do you mean, that in case we need to acuire/extend the lock with another redlock client, we have to pass also the original lock_info, otherwise we get the wrong results?

Correct.

>  it seems like SHA1 is already deprecated for signatures.

Redis `EVALSHA` applies SHA1 on the Lua script to generate the key and cache the script. The gem relies on [`Digest::SHA1`](https://ruby-doc.org/stdlib-2.7.0/libdoc/digest/rdoc/Digest/SHA1.html) to leverage that caching:

https://github.com/leandromoreira/redlock-rb/blob/fc3c5b1557a906b9651d575138c62a1da3086f8e/lib/redlock/scripts.rb#L26-L33

Reply:
> After reading more about https://redis.io/docs/manual/patterns/distributed-locks, I see a few import points of attention:
> 
>     * Under no circumstances concurrent threads (e.g. concurrent sidekiq jobs of the same process) can share a same redlock client as the value of the token that ensures the lock 100% depends on it.

I'm not sure this is true, the lock depends on the `value`, but `value` is different every time `lock` is called, even from the same client. See the [example](https://github.com/leandromoreira/redlock-rb#usage-example) en the README. Or the `lock` method [source](https://github.com/leandromoreira/redlock-rb/blob/main/lib/redlock/client.rb#L276). The value will be different on each call unless you explicitly send the `extend: { value: … }` parameter to `lock`. In order to unlock, you need to pass the `lock_info` object to the `unlock` [method](https://github.com/leandromoreira/redlock-rb/blob/main/lib/redlock/client.rb#L104). So IMO what can't really be shared is the `lock_info` object.

>     * If a lock has to be carried across the stack, rely on (and honour!) the `extend: { value: … }` options of `#lock` when instantiating the redlock client. I don't think that's the case as methods should be concise and self-contained.
> 

From what I wrote above, I think we should never send `extend: { value: … }` to `lock`. The library handles the value correctly. If we need to move the lock to somewhere else, what we need is to move the `lock_info` object. And in general I would stick to:

1 lock_info <-> 1 thread

Reply:
Good information. We don't even care about `lock_info` in current implementation. We can cache the client in a class variable so we don't recreate it every time. But this would be a minor optimization. So mayb we do once we see that new locking will fix/improve the situation.